source/geod24/concurrency.d

```diff
@@ -1006,6 +1007,601 @@ protected:
+            synchronized (this.lock)
+                return this.buffer.length;
+        }
+
+        /***********************************************************************
+
+            An aggrate of neccessary information to block a Fiber and record
+            their request
+
+        ***********************************************************************/
+
+        private class ChannelQueueEntry
+        {
+            /// Semaphore blocking the `Fiber`
+            FiberBinarySemaphore sem;
+
+            /// Pointer to the variable that we will read to/from
+            T* pVal;
+
+            /// Result of the blocking read/write call
+            bool success = true;
+
+            /// State of the select call that this entry was created for
+            SelectState select_state;
+
+            /// Id of the select call that this entry was created for
+            uint sel_id;
+
+            this (T* pVal, SelectState select_state = null, uint sel_id = 0)
+            {
+                this.pVal = pVal;
+                this.select_state = select_state;
+                this.sel_id = sel_id;
+
+                if (this.select_state)
+                    this.sem = this.select_state.sem;
+                else
+                    this.sem = new FiberBinarySemaphore;
+            }
+
+            /***********************************************************************
+
+                Terminate a `ChannelQueueEntry` by waking up the blocked Fiber
+                and reporting the failure
+
+                This is called on all the `ChannelQueueEntry` instances still in
+                the wait queues when Channel is closed
+
+            ***********************************************************************/
+
+            void terminate()
+            {
+                this.success = false;
+                if (!this.select_state || this.select_state.tryConsume(this.sel_id, this.success))
+                    this.sem.notify();
+            }
+        }
+
+        /***********************************************************************
+
+            Create and enqueue a `ChannelQueueEntry` to the given entryq
+
+            Assumes that the Channel is locked, unlocks it before returning
+
+            Param:
+                entryq = Queue to append the new ChannelQueueEntry
+                pVal =  Pointer to the message buffer
+                sel_state = SelectState object to associate with the
+                            newly created ChannelQueueEntry
+                sel_id = id of the select call creating the new ChannelQueueEntry
+            Return:
+                newly created ChannelQueueEntry
+
+        ***********************************************************************/
+
+        private ChannelQueueEntry enqueueEntry(ref ChannelQueueEntry[] entryq, T* pVal, SelectState sel_state = null, uint sel_id = 0)
+        {
+            ChannelQueueEntry q_ent = new ChannelQueueEntry(pVal, sel_state, sel_id);
```

Comment:
Why does a `ChannelQueueEntry` need to be a class? It seems like a lot of overhead. It might be better to just use a struct and avoid unnecessary GC allocations.

Reply:
I needed it to be a heap-allocated-reference-type and went for a class.
Would this function do any allocations other than the one we do explicitly here? I assumed that only the reference of the `q_ent` gets passed around. Wouldn't a heap allocated struct still use GC?

Reply:
Ah if you need reference semantics, it's ok to keep it as a class. I didn't realize that was the use-case.

Reply:
> Wouldn't a heap allocated struct still use GC?

Yes. I wrongly assumed we can use value semantics here.